### ISSUE TITLE
Better error messaging for when property assignments fail due to intersection types

### ISSUE THREAD

**Issue by jeremy-deutsch** [microsoft/TypeScript#42788]
# Bug Report

### 🔎 Search Terms

`error messaging`
`not assignable to type 'never'`

### 🕗 Version & Regression Information

This isn't an error, just a confusing error message for a corner case, so it's been in TypeScript for a while.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.1.3#code/JYOwLgpgTgZghgYwgAgArQM4HsTIN4BQyyIcAthAFzIZhSgDmA3EcnA1SQK5kBG0LAL4ECMLiARhgOZAiwAHAJ7oo2EAAos9BqDgAbairUAaWQsUAVLIcw4AlPlYwtydXJC1kAawiLkWGGQAbQAiUgoQ0xD2CBCAXTYMMw8wB0JiYjklKyCfRQSAXn9tXT1c3zjWYWFRcUlpXCzlOCgpfSMcTRLSAzQWtr0AHg6QAD5TJqtDfuB9YdsxtKcXNxxPPP9A0PDYqJj4xOTaJYyzbKxy-OQirWAdHsvK4mqCIA)

### 💻 Code

```ts
interface Person {
  name: string;
  age: number;
}

function copyPerson(original: Person, copyTo: Person) {
  for (const key of ["name", "age"] as const) {
    // Type 'string | number' is not assignable to type 'never'.
    //   Type 'string' is not assignable to type 'never'. (2322)
    copyTo[key] = original[key]; // error on copyTo[key]
  }
}

function copyPartialPerson(original: Partial<Person>, copyTo: Partial<Person>) {
  for (const key of ["name", "age"] as const) {
    // Type 'string | number | undefined' is not assignable to type 'undefined'.
    //   Type 'string' is not assignable to type 'undefined'. (2322)
    copyTo[key] = original[key]; // error on copyTo[key]
  }
}

```

### 🙁 Actual behavior

The error messages shown in the code above.

### 🙂 Expected behavior

An error message indicating that the right-hand side of the assignment must be assignable to the intersection of all possible types of the left-hand side of the assignment. The intersection part is specifically what seems to be hard to intuit about this error the first time someone encounters it. The `never` and `undefined` types being assigned to in the examples above seem to come out of nowhere.

One option is to add another level of messaging to the error, e.g.
```
Type 'string | number' is not assignable to type 'string & number'.
  Type 'string | number' is not assignable to type 'never'.
    Type 'string' is not assignable to type 'never'.
```

```
Type 'string | number | undefined' is not assignable to type '(string | undefined) & (number | undefined)'.
  Type 'string | number | undefined' is not assignable to type 'undefined'.
    Type 'string' is not assignable to type 'undefined'.
```

Another option is to special-case assignments that assign to `$expr1[$expr2]`, and have an error message like `Cannot assign type $assigneeType to $expr1[$expr2] for all values of $expr2.` e.g.

```
Cannot assign type 'string | number' to 'copyTo[key]' for all possible values of 'key'.
  Type 'string | number' is not assignable to type 'never'.
    Type 'string' is not assignable to type 'never'.
```

```
Cannot assign type 'string | number | undefined' to 'copyTo[key]' for all possible values of 'key'.
  Type 'string | number | undefined' is not assignable to type 'undefined'.
    Type 'string' is not assignable to type 'undefined'.
```

That feels nicer, but we wouldn't want to have that error message any time a bad assignment is made to an index access. To only show the message at the appropriate time, a second check might be necessary, where the type-checker checks if the right-hand side would be assignable to the "read" type of the left-hand side. As in, if the following code would work:
```
let val = leftHandSideExpression;
val = rightHandSideExpression;
```
Then we can guess that the user is probably encountering this specific issue.

**Comment by RyanCavanaugh** [comment]
I agree with your overall analysis of this being a bad error experience. My diagnosis of this is a little different.

When we have a form
```ts
o[k] = e
```
where `k` is a union that produces a disjoint (i.e. intersecting-to-never) set of properties from `o`, then *the type of `e` is irrelevant* and we can issue a more specialized error message:
```
o[k] = e
~~~~
This assignment can write to multiple properties, but no type exists that can safely assign to all of them
```
At a minimum this generates a noncephrase that leads to a good StackOverflow post

**Comment by RyanCavanaugh** [comment]
Thinking about it more, we should do some "back up and try again"-style elaboration here when the LHS is `o[k]` and `k` is a union -- we can complain specifically about the first `k` that isn't legally assigned from `e`.

Finally, we also should really investigate special casing `t[k1] = s[k2]` where `k1` and `k2` are the exact same identifier, since that's really the root cause.

**Comment by BobbieGoede** [comment]
I agree the error message is confusing, after playing around with your example I'm not sure why the following changes returns no errors. I understand that the types are narrowed by checking the keys, but isn't it essentially the same as `copyTo[key] = original[key]`?.

[Playground link](https://www.typescriptlang.org/play?ts=4.1.3#code/JYOwLgpgTgZghgYwgAgArQM4HsTIN4BQyyIcAthAFzIZhSgDmA3EcnA1SQK5kBG0LAL4ECMLiARhgOZAiwAHAJ7oo2EAAos9BqDgAbairUAaWQsUAVLIcw4AlPlYwtydXJC1kAawiLkWGGQAbQAiUgoQ0xD2CBCAXTYMMw8wB0JiYjklKyCfRQSAXn9tXT1c3zjWYWFRcUlpXCzlOCgpfSMcTRLSAzQWtr0AHg6QAD5TJqtDfuB9YdsxtKcXNxxPPP9A0PDYqJj4xOTaJYyzbKxy-OQirWAdHsvK4mqRMQkpGSbUGfaFgGlfBAACZdO6laatWZDEbjM6Wax9SFzGEnZDOKCudzrXybYJhci7ZDRDgHOBJLGpRynYAwdQbAoMok7EJ2dKnOE5PKFYpgh5c1jPAXIGl0nEMorE2KsoWZcycirXHn3fSPIXCQXCIA)

```ts
function copyPartialPersonKeyed(original: Partial<Person>, copyTo: Partial<Person>) {
  for (const key of ["name", "age"] as const) {
    if(key === "name"){
      copyTo[key] = original[key]
    }
    if(key === "age"){
      copyTo[key] = original[key]
    }
  }
}
```

**Comment by RyanCavanaugh** [comment]
TypeScript doesn't evaluate the behavior of unions by combinatorially exploding them into all the possible scenarios -- doing this can improve completeness, but would come at a catastrophic cost to performance
